### PR TITLE
Core: Hoist 'control.options', validate them in core and introduce 'control.labels'

### PR DESCRIPTION
**Pull request by ghengeveld**
Issue: #13888

## What I did

This gracefully replaces `control.options` with top-level `options` on argTypes. Top-level `options` are validated by the story store rather than the Controls addon. Contrary to `control.options`, `options` only accepts an array of primitives. Labels can be customized using `control.labels`.

```js
argTypes: {
  foo: {
    options: [true, false, 42],
    control: {
      type: 'select',
      labels: {
        true: 'Yes',
        false: 'No'
      }
    }
  }
}
```

## How to test

- Is this testable with Jest or Chromatic screenshots? No
- Does this need a new example in the kitchen sink apps? No
- Does this need an update to the documentation? Yes

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: lib/client-api/src/args.ts [hunk 1]
```diff
@@ -50,3 +53,42 @@ export const combineArgs = (value: any, update: any): Args => {
...
+        Invalid argType: '${key}.options' should be an array.
+
+        More info: https://storybook.js.org/docs/react/api/argtypes
+      `);
+      acc[key] = args[key];
+      return acc;
+    }
+
+    if (options.some((opt) => opt && ['object', 'function'].includes(typeof opt))) {
+      once.warn(dedent`
```
Comment (shilman) [inline lib/client-api/src/args.ts:75]: Nice, I like this pattern better than `util-deprecate` 💯

**Review by shilman — CHANGES_REQUESTED** [review]
This looks fantastic. We need to update the controls documentation also!

File: lib/client-api/src/args.ts (hunk 1 above)
Comment (shilman) [reply lib/client-api/src/args.ts:75]: Are we using this `once.warn` for any other purposes? Would be nice to be able to globally search for something with every major release and get a list of all the stuff we have to strip out of the code.

Comment (ghengeveld) [reply lib/client-api/src/args.ts:75]: Look for the word "deprecate" ?

**Review by tmeasday — APPROVED** [review]
Looks great @ghengeveld

File: docs/snippets/common/component-story-custom-args-mapping.js.mdx [hunk 2]
```diff
@@ -0,0 +1,27 @@
+```js
+// YourComponent.stories.js
+
+import startCase from 'lodash/startCase';
+
+import { YourComponent } from './your-component';
+import * as icons from './icons';
```
Comment (shilman) [inline docs/snippets/common/component-story-custom-args-mapping.js.mdx:7]: We should show the explicit value here and not leave it up to the reader's imagination

File: docs/essentials/controls.md [hunk 3]
```diff
@@ -147,35 +151,20 @@ As they can be complex cases:
...
-<CodeSnippets
-  paths={[
-    'common/component-story-custom-args-complex-object.js.mdx',
-  ]}
-/>
-
-<!-- prettier-ignore-end -->
-
-Or even with certain types of elements, such as icons:
+Unless you need the flexibility of a function, an easier way to map primitives to complex values before rendering is to define a `mapping`. Additionally, you can specify `control.labels` to configure custom labels for your checkbox, radio or select input.
```
Comment (shilman) [inline docs/essentials/controls.md:154]: what happens if you need the flexibility of a function? we're probably better off not mentioning this at all?

File: lib/client-api/src/inferControls.ts [hunk 4]
```diff
@@ -10,51 +10,50 @@ type ControlsMatchers = {
...
   switch (type.name) {
     case 'array': {
       const { value } = type;
       if (value?.name && ['object', 'other'].includes(value.name)) {
         return {
-          type: 'object',
-          validator: (obj: any) => Array.isArray(obj),
+          control: {
+            type: 'object',
+            validator: (obj: any) => Array.isArray(obj),
```
Comment (shilman) [inline lib/client-api/src/inferControls.ts:35]: is `validator` documented?

**Review by shilman — CHANGES_REQUESTED** [review]
Looks great to me, modulo a few minor suggestions.

I'm also wondering whether we should have a link from the controls docs back into `MIGRATION.md` for existing controls users who read the updated docs and are confused?

@jonniebigodes can you take a look?

File: docs/snippets/common/component-story-custom-args-mapping.js.mdx (hunk 2 above)
Comment (ghengeveld) [reply docs/snippets/common/component-story-custom-args-mapping.js.mdx:7]: Done

File: docs/essentials/controls.md (hunk 3 above)
Comment (ghengeveld) [reply docs/essentials/controls.md:154]: Should we omit the example with the Template altogether then? I would be totally fine with that.

File: lib/client-api/src/inferControls.ts (hunk 4 above)
Comment (shilman) [reply lib/client-api/src/inferControls.ts:35]: LOL almost zero memory of this 😭 

Per our slack discussion, let's get rid of validators and also use the JSON editor UI for **ALL** arrays moving forward

File: lib/client-api/src/args.test.ts [hunk 5]
```diff
@@ -117,3 +124,43 @@ describe('combineArgs', () => {
...
+    });
+  });
+
+  it('includes arg if no options are specified', () => {
+    expect(validateOptions({ a: 1 }, { a: {} })).toStrictEqual({ a: 1 });
+  });
+
+  it('ignores options and warns if options is not an array', () => {
+    expect(validateOptions({ a: 1 }, { a: { options: { 2: 'two' } } })).toStrictEqual({ a: 1 });
+    expect(once.warn).toHaveBeenCalledWith(
```
Comment (ghengeveld) [inline lib/client-api/src/args.test.ts:152]: Should be error

File: lib/client-api/src/args.test.ts [hunk 6]
```diff
@@ -117,3 +124,43 @@ describe('combineArgs', () => {
...
+  });
+
+  it('warns if options contains non-primitive values', () => {
+    validateOptions({ a: { one: 1 } }, { a: { options: [{ one: 1 }, { two: 2 }] } });
+    expect(once.warn).toHaveBeenCalledWith(
+      expect.stringContaining("Invalid argType: 'a.options' should only contain primitives")
+    );
+    expect(once.warn).toHaveBeenCalledWith(
+      "Received illegal value for 'a'. Supported options: [object Object], [object Object]"
+    );
```
Comment (ghengeveld) [inline lib/client-api/src/args.test.ts:164]: Just act as if options wasnt specified at all, when options isn't valid.

File: docs/essentials/controls.md (hunk 3 above)
Comment (shilman) [reply docs/essentials/controls.md:154]: I mis-read this paragraph the first time around and thought you were referring to the mapping function which we decided not to implement. I'm OK with this as-is.

**Review by shilman — APPROVED** [review]
Looking great!

**Review by ndelangen — APPROVED** [review]

**Pull request merged**